src/t_hash.c

```diff
@@ -7,27 +7,456 @@
+                         db->dict. Will be used from active-expiration flow
+                         for notification and deletion of the object, if
+                         needed. */
+    void *lp;         /* listpack that contains 'key-value-ttl' tuples in it. */
+} listpackTTL;
+
+
+static struct listpackTTL *listpackTTLCreate(size_t lpCap) {
+    listpackTTL *lpt = zcalloc(sizeof(*lpt));
+    lpt->lp = lpNew(lpCap);
+    lpt->meta.trash = 1;
+    lpt->key = NULL;
+    return lpt;
+}
+
+static void listpackTTLFree(listpackTTL *lpt) {
+    lpFree(lpt->lp);
+    zfree(lpt);
+}
+
+/* Returns 1 if listpackTTL contains no fields or all the fields are expired. */
+static int listpackTTLIsEmpty(const robj *o) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_TTL);
+
+    unsigned char *fptr, *s;
+    listpackTTL *lpt = o->ptr;
+
+    if (lpt->meta.trash)
+        return lpLength(lpt->lp) == 0;
+
+    /* If time of next hash-field to expire is greater than current time, then not empty */
+    if (ebGetMetaExpTime(&lpt->meta) >= (uint64_t)commandTimeSnapshot())
+        return 0;
+
+    fptr = lpFirst(lpt->lp);
+    while (fptr != NULL) {
+        long long val;
+
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+
+        s = lpGetValue(fptr, NULL, &val);
+        serverAssert(!s);
+
+        if (!hashTypeListpackIsExpired(val))
+            return 0;
+
+        fptr = lpNext(lpt->lp, fptr);
+    }
+
+    return 1;
+}
+
+/* Returns number of expired fields. */
+static uint64_t listpackTTLExpireDryRun(const robj *o) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_TTL);
+
+    uint64_t expired = 0;
+    unsigned char *fptr, *s;
+    listpackTTL *lpt = o->ptr;
+
+    fptr = lpFirst(lpt->lp);
+    while (fptr != NULL) {
+        long long val;
+
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+
+        s = lpGetValue(fptr, NULL, &val);
+        serverAssert(!s);
+
+        if (hashTypeListpackIsExpired(val))
+            expired++;
+
+        fptr = lpNext(lpt->lp, fptr);
+    }
+
+    return expired;
+}
+
+/* Returns the expiration time of the item with the nearest expiration. */
+static uint64_t listpackTTLGetMinExpire(robj *o) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_TTL);
+
+    uint64_t minExpire = EB_EXPIRE_TIME_INVALID;
+    unsigned char *fptr, *s;
+    listpackTTL *lpt = o->ptr;
+
+    fptr = lpFirst(lpt->lp);
+    while (fptr != NULL) {
+        long long val;
+
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+        fptr = lpNext(lpt->lp, fptr);
+        serverAssert(fptr);
+
+        s = lpGetValue(fptr, NULL, &val);
+        serverAssert(!s);
+
+        if (val != HASH_LP_NO_TTL && (uint64_t) val < minExpire)
+            minExpire = val;
+
+        fptr = lpNext(lpt->lp, fptr);
```

Comment:
sun: here could be a bottleneck, maybe we can add lpForEach with a callback method and use lpSkip to traverse lp internally, it will be much faster than three lpNext().